Mathlib/Data/Nat/Choose/Lucas.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2023 Gareth Ma. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gareth Ma
+-/
+import Mathlib.Data.ZMod.Basic
+import Mathlib.RingTheory.Polynomial.Basic
+
+/-!
+# Lucas's theorem
+
+This file contains a proof of [Lucas's theorem](https://www.wikiwand.com/en/Lucas%27s_theorem) about
+binomial coefficients, which says that for primes `p`, `n` choose `k` is congruent to product of
+`n_i` choose `k_i` modulo `p`, where `n_i` and `k_i` are the base-`p` digits of `n` and `k`,
+respectively.
+-/
+
+open Finset hiding choose
+
+open Nat BigOperators Polynomial
+
+namespace Choose
+
+variable {n k p : ℕ} [Fact p.Prime]
+
+/-- **Lucas's Theorem**: For primes `p`, `choose n k` is congruent to `choose (n % p) (k % p) *
+  choose (n / p) (k / p)` modulo `p`. -/
+theorem lucas_theorem : choose n k ≡ choose (n % p) (k % p) * choose (n / p) (k / p) [ZMOD p] := by
+  have decompose : ((X : (ZMod p)[X]) + 1) ^ n = (X + 1) ^ (n % p) * (X ^ p + 1) ^ (n / p) := by
+    nth_rw 1 [← mod_add_div n p, pow_add, pow_mul, add_pow_char (ZMod p)[X] (p := p), one_pow]
```

Comment:
Couldn't this be an independent lemma?

Reply:
I don't know the proof here, but the question above prompts asking if you can decompose this proof into further pieces!

Reply:
This lemma is saying that $(X + 1)^n \equiv (X + 1)^{n \pmod{p}}(X^p + 1)^{\lfloor n / p \rfloor} \pmod{p}$ which follows because $(X + 1)^n \equiv (X + 1)^{n \pmod{p} + \lfloor n / p \rfloor \cdot p} \pmod{p}$ then use Freshman's dream. I will isolate it out into `Mathlib/Algebra/CharP/Basic.lean` then :D